src/util-sysfs.c

```diff
@@ -0,0 +1,69 @@
+/* Copyright (C) 2011-2022 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Richard McConnell <richard_mcconnell@rapid7.com>
+ *
+ * Sysfs utility file
+ */
+
+#include "util-sysfs.h"
+
+#define SYSFS_MAX_FILENAME_LEN (SYSFS_MAX_FILENAME_SIZE + 5)
+
+TmEcode SysFsWriteValue(const char *path, int64_t value)
+{
+#if defined(__linux__)
+    struct stat stats;
+    char fname[SYSFS_MAX_FILENAME_LEN] = "/sys/";
+
+    if (!path || strlen(path) > SYSFS_MAX_FILENAME_SIZE) {
+        SCLogWarning(SC_ERR_ARG_LEN_LONG, "File path too long, max allowed: %d",
+                SYSFS_MAX_FILENAME_SIZE);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    strlcat(fname, path, sizeof(fname));
+
+    if (stat(fname, &stats) != 0) {
+        SCLogError(SC_ERR_FOPEN, "Could not gather stats for: %s", fname);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (S_ISREG(stats.st_mode) == 0) {
+        SCLogError(SC_ERR_FOPEN, "File: %s is not a regular file", fname);
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    FILE *fp = fopen(fname, "w");
```

Comment:
Interesting, think this is the first time CodeQL flagged a PR.

For some ideas on how to avoid the toctou, see e.g. 875eb58fb05a9075b3020e69cf8940d1d2fb6ad2

Reply:
@victorjulien I have developed another version and wanted to hear your thoughts:
``` C
if defined(__linux__)
    char fname[SYSFS_MAX_FILENAME_LEN] = "/sys/";
    char sentence[64];

    if (!path || strlen(path) > SYSFS_MAX_FILENAME_SIZE) {
        SCLogWarning(SC_ERR_ARG_LEN_LONG, "File path too long, max allowed: %d",
                SYSFS_MAX_FILENAME_SIZE);
        SCReturnInt(TM_ECODE_FAILED);
    }   

    strlcat(fname, path, sizeof(fname));

    /* File must be present and process have correct capabilities to open */                                                                                                                                                                                                                                                  
    int fd = open(fname, O_WRONLY);
    if (fd < 0) {
        SCLogError(SC_ERR_FOPEN, "Could not open file: %s", fname);
        SCReturnInt(TM_ECODE_FAILED);
    }   

    snprintf(sentence, sizeof(sentence), "%ld", value);
    ssize_t len = strlen(sentence);

    if (write(fd, sentence, len) != len) {
        SCLogError(SC_ERR_FWRITE, "Could not write to file: %s", fname);
        close(fd);
        SCReturnInt(TM_ECODE_FAILED);
    }   
    close(fd);
#endif /* __LINUX__ */

```
With this I am checking the presence of the file and the process capabilities during the ``open()`` and it shall fail if either are false. If you are happy with this is it best I remove the v3 PR or create a v4? As v3 has not yet been reviewed I am confused the best action to take here.

Reply:
The code above seems reasonable to me. Can you include that in a v4 after further review?

Reply:
Yeah, sure can. Thanks :)